RecoEgamma/EgammaElectronAlgos/interface/ConversionFinder.h

```diff
@@ -0,0 +1,101 @@
+   GSF track). Calculate the dist, dcot, point of conversion and the
+   radius of conversion for this pair and fill the ConversionInfo
+*/
+
+namespace egamma::conv {
+
+  struct ConversionInfo {
+    const float dist;
+    const float dcot;
+    const float radiusOfConversion;
+    // if the partner track is found in the  GSF track collection,
+    // this is a ref to the GSF partner track
+    const std::optional<int> conversionPartnerCtfTkIdx;
+    // if the partner track is found in the  CTF track collection,
+    // this is a ref to the CTF partner track
+    const std::optional<int> conversionPartnerGsfTkIdx;
+    const int deltaMissingHits;
+    const int flag;
+
+    // flag 0: Partner track found in the CTF collection using the electron's CTF track
+    // flag 1: Partner track found in the CTF collection using the electron's GSF track
+    // flag 2: Partner track found in the GSF collection using the electron's CTF track
+    // flag 3: Partner track found in the GSF collection using the electron's GSF track
+  };
+
+  namespace col {
+    SOA_DECLARE_COLUMN(PtError, float, "ptError");
+    SOA_DECLARE_COLUMN(D0, float, "d0");
+
+    SOA_DECLARE_COLUMN(NumberOfValidHits, int, "numberOfValidHits");
+    SOA_DECLARE_COLUMN(MissingInnerHits, int, "missingInnerHits");
+    SOA_DECLARE_COLUMN(Charge, int, "charge");
+  }  // namespace col
+
+  using TrackSpecificColumns =
+      std::tuple<col::PtError, col::MissingInnerHits, col::NumberOfValidHits, col::Charge, col::D0>;
+  using TrackTable = edm::soa::AddColumns<edm::soa::PtEtaPhiThetaTable, TrackSpecificColumns>::type;
+  using TrackTableView = edm::soa::ViewFromTable_t<TrackTable>;
+  using TrackRowView = TrackTable::const_iterator::value_type;
+
+  template <class Object>
+  TrackTable makeTrackTable(std::vector<Object> const& objects) {
+    return {
+        objects,
+        edm::soa::column_fillers(col::D0::filler([](Object const& x) { return x.d0(); }),
+                                 edm::soa::col::Pt::filler([](Object const& x) { return x.pt(); }),
+                                 edm::soa::col::Theta::filler([](Object const& x) { return x.theta(); }),
+                                 col::PtError::filler([](Object const& x) { return x.ptError(); }),
+                                 col::MissingInnerHits::filler([](Object const& x) {
+                                   return x.hitPattern().numberOfLostHits(reco::HitPattern::MISSING_INNER_HITS);
+                                 }),
+                                 col::NumberOfValidHits::filler([](Object const& x) { return x.numberOfValidHits(); }),
+                                 col::Charge::filler([](Object const& x) { return x.charge(); }),
+                                 edm::soa::col::Eta::filler([](Object const& x) { return x.eta(); }),
+                                 edm::soa::col::Phi::filler([](Object const& x) { return x.phi(); }))};
+  }
```

Comment:
When KinematicTables was introduced, I was hoping/assuming that some reuse and perhaps multiplexing can be practical in some common-looking way.

Looking at this part, it doesn't look all that coherent, but perhaps that's supposed to be tolerated ;)

Looking more carefully
```
#define SOA_DECLARE_COLUMN(_ClassName_, _Type_, _String_)             \
  struct _ClassName_ : public edm::soa::Column<_Type_, _ClassName_> { \
```
e.g. for PtError will lead to struct ```egamma::conv::col::PtError``` inheriting from ```edm::soa::Column<float, PtError>```. If I parsed it right, then the base ```edm::soa::Column<float, PtError>``` is not a good result because it gives a very high chance for symbol conflicts; we should have it in the local namespace instead or move this out of here to a common place.
Would ```SOA_DECLARE_COLUMN(egamma::conv::col::PtError, float, "PtError")```  work (the macro call will need to be in a global namespace)?

I'm a somewhat concerned that the TrackTable implementation depends on ```PtEtaPhiThetaTable``` (which comes from CommonTools/Utils/interface/KinematicTables.h and does not have a make\*Table method) , while the definition of how ```PtEtaPhiThetaTable``` is filled is fully defined here in a local code.
Perhaps this means that ways of making the tables should not really be in the place where the tables are declared; OTOH, would it be practical to have a ```makePtEtaPhiThetaTable``` defined in KinematicTables.h and reused here?

If ways of filling the tables are not reusable, and if the tables themselves are somewhat trivial, and if they are needed only in some local package, is there really a benefit in trying to define/use a common location?
My thought would be to keep it fully local for the algorithm to avoid the symbol or conflicting definition mess.


BTW, it looks like in this PR tanTheta = pt/pz should be used instead of theta (which is computed only to then compute a tan of it).

Reply:
Thanks for sharing your thoughts! No, this is not supposed to be tolerated, I was just kindly asking for using SOA for all the track information to be tolerated, to not having to change the Track classes or expanding the interface of `findConversion` to take both a track table and the track collection.

So I understand that one hand you want to avoid name collisions, and on the other hand you want to organize the SoA layouts in a better way. I guess we can avoid name collisions by something like
```C++
namespace egamma::conv::col {
    PtError;
    ....
}
SOA_DECLARE_COLUMN(egamma::conv::col::PtError, float, "PtError");.
....
```
I think it's not possible to directly create a struct in a namespace like `struct foo::A {}` without such a forward declaration. So the name collisions can be avoided at the cost of higher verbosity, which is the bare minimum of what I should do.

I absolutely share your opinions about everything else you elaborate, and indeed the fact that these fillers can't be factorized while the tables can is a good starting point to ponder more about FWCore/SOA. To me it would even make sense to declare the fillers as static functions in the column struct itself, for example.

Maybe my problem here is this: I am convinced that whatever we would come up with here to make us all happy will require changes in `FWCore/SOA`, which doesn't really belong to this speedup PR. So maybe we should just keep things local for now while avoiding namespace collisions and then at some later point there will be a PR along the lines of "extending SOA interface after lessons learned using it".

Reply:
> Maybe my problem here is this: I am convinced that whatever we would come up with here to make us all happy will require changes in `FWCore/SOA`, which doesn't really belong to this speedup PR. So maybe we should just keep things local for now while avoiding namespace collisions and then at some later point there will be a PR along the lines of "extending SOA interface after lessons learned from using it".

If there are concerns about timeline, I think that there is no rush.
If `FWCore/SOA` is in question, perhaps we should have a related discussion in the core meeting.
@makortel what do you think?

Reply:
That would be good! My concern is not really the timescale, but the lack of data. `FWCore/SOA` is so far only used in my previous optimization PR and this one, so it's challenging do decide on best practices and potential new interfaces now.

Reply:
> If `FWCore/SOA` is in question, perhaps we should have a related discussion in the core meeting.

We can certainly discuss about the "lessons learned from using `FWCore/SOA`" in an upcoming core meeting.



> Looking more carefully
> 
> ```
> #define SOA_DECLARE_COLUMN(_ClassName_, _Type_, _String_)             \
>   struct _ClassName_ : public edm::soa::Column<_Type_, _ClassName_> { \
> ```
> 
> e.g. for PtError will lead to struct `egamma::conv::col::PtError` inheriting from `edm::soa::Column<float, PtError>`.

No. Plain
```cpp
SOA_DECLARE_COLUMN(PtError, float, ...);
```
would lead to a struct `PtError` inheriting from `edm::soa::Column<float, PtError>`. But
```cpp
namespace egamm::conv::col {
  SOA_DECLARE_COLUMN(PtError, float, ...);
}
```
leads to a struct `egamma::conv::col::PtError` that inherits from `edm::soa::Column<float, egamma::conv::col::PtError>`.

Reply:
> When KinematicTables was introduced, I was hoping/assuming that some reuse and perhaps multiplexing can be practical in some common-looking way.

Assuming I understood the comment correctly, how about overloading `value_for_column()` instead of using `edm::soa::column_fillers()`?
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideStructureOfArrays#Constructors

Reply:
> But
> 
> ```c++
> namespace egamm::conv::col {
>   SOA_DECLARE_COLUMN(PtError, float, ...);
> }
> ```
> 
> leads to a struct `egamma::conv::col::PtError` that inherits from `edm::soa::Column<float, egamma::conv::col::PtError>`.

Nice. Thanks for correcting my initial reading.

Reply:
Hi! Sorry I missed the conversation, I just discovered `value_for_column` accidentally myself while trying to implement the same functionality. That is awesome! I think it solves most our concerns on how these table layouts factorize. New commit n a second.

Reply:
Probably you'll frown upon the new `SOA_DECLARE_DEFAULT` macro but it's a start :)

Reply:
The igprof reports are in the PR description now too

Reply:
> Probably you'll frown upon the new `SOA_DECLARE_DEFAULT` macro but it's a start :)

right, http://cms-sw.github.io/cms_coding_rules.html#4--technical-coding-rules-1
> 4.15 Do not use function-like macros.

Reply:
On the other hand the macro is used to define a function, not like a function.

Reply:
Indeed, I was referring more to the `return x._Expression_;` hack which is a bit on the far end of the flexibility vs. avoiding verbosity spectrum.